web/source/osk/visualKeyboard.ts

```diff
@@ -582,32 +585,25 @@ namespace com.keyman.osk {
      **/
     release: (e: TouchEvent) => void = function(this: VisualKeyboard, e: TouchEvent) {
       // Prevent incorrect multi-touch behaviour if native or device popup visible
-      var sk = document.getElementById('kmw-popup-keys'), t = this.currentTarget;
+      var t = this.currentTarget;
 
       // Clear repeated backspace if active, preventing 'sticky' behavior.
       this.cancelDelete();
 
-      if((sk && sk.style.visibility == 'visible')) {
+      if((this.subkeyGesture && this.subkeyGesture.isVisible())) {
         // Ignore release if a multiple touch
         if(e.touches.length > 0) {
           return;
         }
 
-        // Cancel (but do not execute) pending key if neither a popup key or the base key
-        if((t == null) || ((t.id.indexOf('popup') < 0) && (t.id != this.popupBaseKey.id))) {
-          this.highlightKey(this.keyPending,false);
-          this.clearPopup();
-          this.keyPending = null;
-          this.touchPending = null;
+        if(this.subkeyGesture instanceof browser.SubkeyPopup) {
+          let subkeyPopup = this.subkeyGesture as browser.SubkeyPopup;
+          subkeyPopup.finalize(e.changedTouches[0]);
         }
-      }
+        this.highlightKey(this.keyPending,false);
+        this.keyPending = null;
+        this.touchPending = null;
 
-      // Only set when embedded in our Android/iOS app.  Signals that the device is handling
-      // subkeys, so we shouldn't allow output for the base key.
-      //
-      // Note that on iOS (at least), this.release() will trigger before kmwembedded.ts's
-      // executePopupKey() function.
-      if(this.popupVisible) {
         return;
       }
```

Comment:
Handles subkey completion.

Note that for the in-browser mode, this was previously fall-through; `this.keyPending` held the currently-selected subkey if it existed, and the rest of the code afterward was responsible for processing the keystroke.  Now, we let the `Promise`s handle that.

The coding pattern also ensures that for the embedded codepath, no base-key release processing may occur before a host app resolves the subkey popup.  Thus, the functionality of the `if(this.popupVisible)` block is maintained.